### PR TITLE
Some snap improvements

### PR DESCRIPTION
**Pull request by superm1**
Hodge podge of snap related improvements to shrink size, and to set up directory structure to carry info from one snap to another.

Some of these should have done when first confined, but better now than never.

Also the directory remote is now accessible from host at `/var/snap/fwupd/share/fwupd/remotes.d/vendor/firmware`

CC @CragW 
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Review by CragW — APPROVED** [review]

**Comment by CragW** [comment]
Before this commit, the remote conff and firmware updates are separated in two different places in the snap world.

For instance adding a new remote requires the conff path to `/root/snap/fwupd/current/etc/fwupd/remotes.d/foo-embargo.conf`, which is backed by `$SNAP_USER_DATA`.

Meanwhile, adding a new foo.cab to be installed from the `vendor-directory` remote would require path to `/var/snap/fwupd/common/foo.cab`, which is backed by `$SNAP_COMMON`.

After this commit, the new location would resepect to `$SNAP_COMMON` for both items in terms of `/var/snap/fwupd/common/`.

**Comment by hughsie** [comment]
My concern here is that we've been telling people on the LVFS[1] to put the remote into /root/snap/fwupd/current/etc/fwupd/remotes.d for years -- I don't think the old path and new path are going to be symlinked, right?

[1] https://fwupd.org/lvfs/metadata/

**Comment by superm1** [comment]
> My concern here is that we've been telling people on the LVFS[1] to put the remote into /root/snap/fwupd/current/etc/fwupd/remotes.d for years -- I don't think the old path and new path are going to be symlinked, right?

They're not going to be symlinked, but how about if we make that another search directory for remotes?

**Comment by hughsie** [comment]
> but how about if we make that another search directory for remotes?

Yes, that works -- we just have to avoid making the ODMs **do anything*.

**Comment by superm1** [comment]
I pushed a change, but not yet tested.  Let me know what you think of this conceptually at least.

1. Don't reproduce `/etc` in `SNAP_COMMON`.  Mark `/etc` as immutable (like how CoreOS would do it)
2. Migrate `remotes.d` from `SNAP_USER_DATA` to `SNAP_COMMON` if it exists and isn't in `SNAP_COMMON`
3. Move vendor directory to `/var/snap/fwupd/common/var/fwupd/remotes.d/vendor/firmware`

I'm not really sure on 3 if that's logical or not.  It's kinda like a super remotes directory then that takes remote conffiles from the host in `/var/snap/fwupd/common/var/fwupd/remotes.d/` or auto-cabs in `/var/snap/fwupd/common/var/fwupd/remotes.d/vendor/firmware`

**Comment by hughsie** [comment]
thb, I don't really understand the intricacies of snap -- and I don't mind how it's set up as long as we can either support old-snap+new-snap locations -- OR -- we have a flag day, migrate everything over when installing the new snap and change the instructions on the LVFS to include the new path.

**Comment by superm1** [comment]
> thb, I don't really understand the intricacies of snap -- and I don't mind how it's set up as long as we can either support old-snap+new-snap locations -- OR -- we have a flag day, migrate everything over when installing the new snap and change the instructions on the LVFS to include the new path.

OK, I've got it setup that we can support both locations now.  I've tested the various flows, and it looks right to me.
It will migrate the files if they were put in the `/root` directory.

After we release 1.9.12, I suggest we update the LVFS instructions to use the new path.  

"Some" day we can drop the migration from `/root` stuff.

**Comment by superm1** [comment]
> After this commit, the new location for both items would resepect to $SNAP_COMMON, e.g.:
> /var/snap/fwupd/common/etc/fwupd/remotes.d/foo-embargo.conf
> /var/snap/fwupd/common/share/fwupd/remotes.d/vendor/firmware/foo.cab

With the changes I did today including the migration the two locations are:
* `/var/snap/fwupd/common/var/lib/fwupd/remotes.d/foo-embargo.conf`
* `/var/snap/fwupd/common/share/fwupd/remotes.d/vendor/firmware/foo.cab`
If a user puts an `embargo.conf` in `/root/snap/fwupd/current/etc/fwupd/remotes.d` it will migrate to `/var/snap/fwupd/common/var/lib/fwupd/remotes.d/embargo.conf` if that doesn't exist.

File: .gitignore [hunk 1]
```diff
@@ -35,3 +35,4 @@ plugins/uefi-dbx/tests/
 .buildconfig
 .ossfuzz
 *.rej
+/snap
```
Comment (hughsie) [inline .gitignore:38]: Ohh?

Comment (superm1) [reply .gitignore:38]: When you build the snap you need to put the YAML file in that directory, but we should never commit that directory

Comment (hughsie) [reply .gitignore:38]: Snap is quite opinionated it seems :)

File: src/fu-engine.c [hunk 2]
```diff
@@ -8726,6 +8726,12 @@ fu_engine_constructed(GObject *obj)
 							    XMLB_MICRO_VERSION);
 		fu_context_add_compile_version(self->ctx, "com.hughsie.libxmlb", version);
 	}
+
+	/* add optional snap version */
+	if (g_getenv("SNAP_REVISION") != NULL)
```
Comment (hughsie) [inline src/fu-engine.c:8731]: you need {} here

**Review by hughsie — DISMISSED** [review]
One nit then LGTM -- not a snap expert by a long shot tho

**Pull request merged**